src/main/java/com/azure/cosmos/kafka/connect/sink/id/strategy/ProvidedInStrategy.java

```diff
@@ -0,0 +1,43 @@
+package com.azure.cosmos.kafka.connect.sink.id.strategy;
+
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.data.Values;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.sink.SinkRecord;
+
+import java.util.Map;
+
+class ProvidedInStrategy extends AbstractIdStrategy {
+    protected enum ProvidedIn {
+        KEY,
+        VALUE
+    }
+
+    private final ProvidedIn where;
+
+    ProvidedInStrategy(ProvidedIn where) {
+        this.where = where;
+    }
+
+    @Override
+    public String generateId(SinkRecord record) {
+        Object value = where == ProvidedIn.KEY ? record.key() : record.value();
+        Schema idSchema = null;
+        Object idValue;
+
+        if (value instanceof Struct) {
+            idSchema = ((Struct) value).schema().field(AbstractIdStrategyConfig.ID).schema();
+            idValue = ((Struct) value).get(AbstractIdStrategyConfig.ID);
+        } else if (value instanceof Map) {
+            idValue = ((Map) value).get(AbstractIdStrategyConfig.ID);
+        } else {
+            throw new ConnectException("Expected " + where + " to be of struct or map type to use ProvidedIn strategy");
+        }
+
+        if (idValue == null) {
+            throw new ConnectException("Cannot find id in " + where + ": " + value);
+        }
+        return Values.convertToString(idSchema, idValue);
```

Comment:
what is the encoding used?

There are certain characters which are not allowed by Cosmos DB in id:
"The following characters are restricted and cannot be used in the Id property: '/', '\\', '?', '#'"
https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.documents.resource.id?view=azure-dotnet

Reply:
Are we able to escape those characters? While we probably won't introduce those special characters as part of the serialization to string, there is no preventing the user from just wanting to specify their id to be "#my\special/id?" . We have connectors that do some sanitization of unallowed characters and basically just replace them with a '_'. We could potentially do this here as well. We'd also need to disallow those characters from being used in the template for TemplateStrategy.